optuna/_gp/optim_mixed.py

```diff
@@ -0,0 +1,267 @@
+    # We use an isotropic kernel, so scaling the gradient will make
+    # the hessian better-conditioned.
+    continuous_param_scale = 1 / np.sqrt(inverse_squared_lengthscales[continuous_params])
+    continuous_scaled_bounds = [(0, 1 / s) for s in continuous_param_scale]
+
+    noncontinuous_params = np.where(steps > 0)[0]
+    noncontinuous_param_choices = [
+        (
+            np.arange(bounds[i, 1])
+            if scale_types[i] == ScaleType.CATEGORICAL
+            else normalize_one_param(
+                param_value=np.arange(bounds[i, 0], bounds[i, 1] + 0.5 * steps[i], steps[i]),
+                scale_type=ScaleType(scale_types[i]),
+                bounds=(bounds[i, 0], bounds[i, 1]),
+                step=steps[i],
+            )
+        )
+        for i in noncontinuous_params
+    ]
+    noncontinuous_paramwise_xtol = [
+        # Enough to discriminate between two choices.
+        np.min(np.diff(choices), initial=np.inf) / 4
+        for choices in noncontinuous_param_choices
+    ]
+
+    normalized_params = initial_normalized_params.copy()
+    fval = float(eval_acqf_no_grad(acqf_params, normalized_params))
+
+    def optimize_continuous() -> bool:
+        nonlocal normalized_params, fval
+
+        old_continuous_params = normalized_params[continuous_params].copy()
+
+        def negfun_continuous_with_grad(x: np.ndarray) -> tuple[float, np.ndarray]:
+            normalized_params[continuous_params] = x * continuous_param_scale
+            (fval, grad) = eval_acqf_with_grad(acqf_params, normalized_params)
+            # Flip sign because scipy minimizes functions.
+            return (-fval, -grad[continuous_params] * continuous_param_scale)
+
+        x_opt, fval_opt, info = so.fmin_l_bfgs_b(
+            func=negfun_continuous_with_grad,
+            x0=normalized_params[continuous_params] / continuous_param_scale,
+            bounds=continuous_scaled_bounds,
+            pgtol=math.sqrt(tol),
+            maxiter=200,
+        )
+
+        if info["warnflag"] == 2:
+            fval_opt = negfun_continuous_with_grad(x_opt)[0]
+
+        if -fval_opt < fval:
+            normalized_params[continuous_params] = old_continuous_params
+            return False
+
+        normalized_params[continuous_params] = x_opt * continuous_param_scale
+        fval = -fval_opt
+
+        return info["nit"] != 0  # True if parameters are updated.
+
+    def optimize_exhaustive_search(param_idx: int, choices: np.ndarray) -> bool:
+        nonlocal normalized_params, fval
+
+        current_choice = normalized_params[param_idx]
+        choices_except_current = choices[choices != normalized_params[param_idx]]
+
+        all_params = np.repeat(normalized_params[None, :], len(choices_except_current), axis=0)
+        all_params[:, param_idx] = choices_except_current
+        fvals = eval_acqf_no_grad(acqf_params, all_params)
+        best_idx = np.argmax(fvals)
+
+        if fvals[best_idx] >= fval:
```

Comment:
Why is it inequality with the equal? (Do we count tie as improvement?)
```suggestion
        if fvals[best_idx] > fval:
```

Reply:
Sorry, that was a mistake (it might fall into infinite loop).